server/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -544,6 +552,26 @@ public ShardStartedClusterStateTaskExecutor(AllocationService allocationService,
             List<ShardRouting> shardRoutingsToBeApplied = new ArrayList<>(tasks.size());
             Set<ShardRouting> seenShardRoutings = new HashSet<>(); // to prevent duplicates
             for (StartedShardEntry task : tasks) {
+                final IndexMetaData indexMetaData = currentState.metaData().index(task.shardId.getIndex());
+                if (indexMetaData == null) {
+                    // tasks that correspond to non-existent indices are marked as successful
+                    logger.debug("{} ignoring shard started task [{}] (unknown index {})", task.shardId, task, task.shardId.getIndex());
+                    builder.success(task);
+                    continue;
+                } else {
+                    if (task.primaryTerm > 0) {
```

Comment:
I think we should only do this check if `matched != null && matched.primary()`. This otherwise relies too much on the current behavior that a primary failover leads to a reinitialization of the current replicas (see RoutingNodes#failShard and RoutingNodes#startShard).

Reply:
Right, we talked about it.